<a href="https://colab.research.google.com/github/connerdandrews/cse380-notebooks/blob/master/07_2_Ponder_and_Prove_Elementary_Number_Theory.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ponder and Prove Elementary Number Theory
#### Due: Saturday, 20 February 2021, 11:59 pm.

## Explore Fermat's Little Theorem Further


Fermat's Little Theorem (FLT) says that if $N$ is prime, then $N$ divides $X^N - X$.

Remember, the contrapositive of the conditional statement in this theorem is that if $N$ **doesn't** divide $X^N - X$ for some $X$, then $N$ **can't** be prime.

Unfortunately, this simple primality test doesn't always work, because it can be fooled by **pseudoprimes**.

For example, $341 = 11 \cdot 31$ is not prime. But $341$ **does** divide $2^{341} - 2$ as verified below:

In [ ]:
((2 ** 341) - 2) % 341

So $341$ is a so-called **base-2 pseudoprime**. What about **base-3**?

In [ ]:
((3 ** 341) - 3) % 341

Check that the result is not zero, therefore $341$ is **not** a **base-3 pseudoprime**.

Are there any other bases that will not fool the FLT test for $341$?

Are there any pseudoprimes that will fool the FLT for **any choice** of base coprime to them?

### The answer is yes.

Your task is the find the first 4-digit **bullet-proof pseudoprime** (**bppp**) and **prove** (yes, **PROVE**) that it will fool the FLT test for every base coprime to it.

Your proof must use all of the following:
1. the definition of coprime,
2. a consequence of coprimality,
3. the factorization of the **bppp**,
4. FLT, and the
5. CRT (Chinese Remainder Theorem).


In [3]:
from math import gcd
from sympy import isprime

def passes_FLT_test_even_though_not_prime(b, n):
  # primes don't count as pseudoprimes
  return not isprime(n) and (b ** n) % n == b

def is_bppp(n):
  bases_coprime_to_n = [b for b in range(2, n) if gcd(b, n) == 1]
  return all(list(map(lambda b: passes_FLT_test_even_though_not_prime(b, n), 
                      bases_coprime_to_n)))

n = 1000
while not is_bppp(n):
  n += 1

n

1105

We now know by calculating that the first bullet proof pseudo-prime is 1105. What follows is a proof that 1105 truly is a bullet proof pseudo-prime. 

1105 is a 4-digit composite number. 

This is true because 1105 reduces into its prime factorization: 5 x 13 x 17

gcd(b, 1105) = 1, gcd(b, 5) = 1, gcd(b, 13), gcd(b, 17)

This means that if b is comprime to 1105 then it is coprime to 5, 13, and 17.

Fermat's Little Theorem tells us that b^n-1) is congruent mod n to 1. 

Using this understanding we can then write the congruences that:
 
$b^{4} \equiv_{5} 1$

$b^{12} \equiv_{13} 1$

$b^{16} \equiv_{17} 1$

## What is True?
Assess yourself on how you did using the checkboxes below. Check a box by putting an 'X' in it only if it is warranted.


# TODO My Report on What I Did and What I Learned

## Fun
Replace these words with your own detailing how you had fun (if you did).

## New
Replace these words with your own detailing what was something new you learned (if anything).

## Meaningful
Replace these words with your own detailing what you achieved that was meaningful or that you can build upon.

## Other
Replace these words with your own describing other topics or sections of your report --- Connections, Collaborator Contributions, or anything else you feel impressed to add.

# TODO What is True?
Click on each warranted checkbox to toggle it to True (or back to False). 

NOTE: *This only works in Colab. If you run it in some other Jupyter notebook client/server environment you may have to change False to True (or vice versa) manually.*

This self-assessment is subject to revision by a grader.

In [ ]:
#@markdown ## What is True about what I did?
#@markdown ### I had fun.
cb00 = False #@param {type:'boolean'}
#@markdown ### I learned something new.
cb01 = False #@param {type:'boolean'}
#@markdown ### I achieved something meaningful, or something I can build upon at a later time.
cb02 = False #@param {type:'boolean'}
#@markdown ## What is true about my report?
#@markdown ### I wrote a sufficient number of well-written sentences.
cb03 = False #@param {type:'boolean'}
#@markdown ### My report is free of mechanical infelicities.
cb04 = False #@param {type:'boolean'}
#@markdown ### I used Grammarly (or something better described in my report) to check for MIs.
cb05 = False #@param {type:'boolean'}
#@markdown ### I reported on any connections I found between these problems and something I already know. 
cb06 = False #@param {type:'boolean'}
#@markdown ### I reported who were and what contribution each of my collaborators made.
cb07 = False #@param {type:'boolean'}
#@markdown ## What is true about my proof?
#@markdown ### It succinctly uses the definition of coprime.
cb08 = False #@param {type:'boolean'}
#@markdown ### It correctly uses the definition of coprime.
cb09 = False #@param {type:'boolean'}
#@markdown ### It succinctly uses a consequence of comprimality
cb10 = False #@param {type:'boolean'}
#@markdown ### It correctly uses a consequence of comprimality
cb11 = False #@param {type:'boolean'}
#@markdown ### It succinctly uses the factorization of the **bppp**,
cb12 = False #@param {type:'boolean'}
#@markdown ### It correctly uses the factorization of the **bppp**,
cb13 = False #@param {type:'boolean'}
#@markdown ### It succinctly uses Fermat's Little Theorem. 
cb14 = False #@param {type:'boolean'}
#@markdown ### It correctly uses Fermat's Little Theorem. 
cb15 = False #@param {type:'boolean'}
#@markdown ### It succinctly uses the Chinese Remainder Theorem.
cb16 = False #@param {type:'boolean'}
#@markdown ### It correctly uses the Chinese Remainder Theorem.
cb17 = False #@param {type:'boolean'}



## DO NOT CHANGE ANYTHING IN THE NEXT CODE CELL!!
### Delete this cell and the following ones before submitting your work.

In [ ]:
points_for_what_I_did = [5]*3
points_for_my_report = [7]*5
points_for_my_proof = [5]*10
points = points_for_what_I_did + points_for_my_report + points_for_my_proof
# cb is short for checkbox
total = sum(map(lambda n, p: p if eval(f'cb{n:02}') else 0,
                range(len(points)), points))             
total

# For graders

In [ ]:
#@markdown ---
number_of_MIs_found = 0 #@param {type: 'slider', min: 0, max: 5}
#@markdown ---
